packages/cactus-common/src/main/typescript/exception/exception-helper.ts

```diff
@@ -0,0 +1,327 @@
+import { RuntimeError } from "run-time-error";
+import stringify from "fast-safe-stringify";
+import { Logger } from "../logging/logger";
+
+export class ExceptionHelper {
+  /**
+   * Fetches the exception-content and submits it to the logger
+   *
+   * @param logger logger created thru the LoggerProvider.getOrCreate()
+   * @param exception as an arbitrary exception of whatever type and format it might be
+   * @param optionalAdditionalContext as optional on top info / context on the exception scenario
+   */
+  public static logException(
+    logger: Logger,
+    exception: unknown,
+    optionalAdditionalContext = "",
+  ): void {
+    const exceptionContent = this.getExceptionContent(
+      exception,
+      optionalAdditionalContext,
+    );
+
+    // log to destination
+    logger.error(...exceptionContent);
+  }
+
+  /**
+   * Creates a new RuntimeError based on the incoming exception and context to preserve the complete hierarchy of exception information
+   *
+   * @param exception as an arbitrary exception of whatever type and format it might be
+   * @param context providing context about the exception scenario
+   * @returns a new RuntimeError created from exception to preserve the complete hierarchy of exception information
+   */
+  public static getNewRuntimeErrorFromException(
+    exception: unknown,
+    context: string,
+  ): RuntimeError {
+    const exceptionMessageInfo = this.getExceptionMessageInfo(exception);
```

Comment:
@m-courtin Could this be shortened so that it checks for `exception` being an `instanceof Error` and if yes, use it as the cause, otherwise coerce it to string with the `getExceptionStack` method? (that way it's just a single if condition but with equivalent output AFAICT)

Reply:
@petermetz: the `getExceptionMessageInfo` is handling 4 different scenarios with 4 different outcomes:

- Scenario 1+2: Handling any exception / error / RuntimeError containing a message property
      - Scenario 1: message property is of string type (so no stringify is needed) and is directly taken as returned message
      - Scenario 2: message property is not of type string and therefore needs to be stringified

- Scenario 3: A plain string only was thrown and therefore the returned message gets a PREFIX attached mentioning that a custom exception was thrown (therefore not all expected properties like stack etc might be available)

- Scenario 4: It is an arbitrary exception not containing a message property. So the whole exception with all the included properties will be stringified and a PREFIX attached mentioning that a custom exception was thrown (therefore not all expected properties like stack etc might be available)

Reply:
@m-courtin Sorry I meant the entire `newRuntimeError` function's code, not the code of the `getExceptionMessageInfo` function which is being invoked on the line I'm commenting on. Apologies for the confusion!

Reply:
@m-courtin Sorry I meant the entire `newRuntimeError` function's code, not the code of the `getExceptionMessageInfo` function which is being invoked on the line I'm commenting on. Apologies for the confusion!

Reply:
@petermetz: The `newRuntimeError` function is containing a distinction in the else code to prevent duplicated parts of information in the returned message. I added some lines of further detail documentation to the function to describe it further.

Reply:
> prevent duplicated parts of information in the returned message

@m-courtin Right now `ExceptionHelper.newRuntimeError("Oops", "Doh").cause` evaluates to `"A CUSTOM EXCEPTION WAS THROWN: Oops - A CUSTOM EXCEPTION WAS THROWN: \"Oops\""` anyway (I know it's the cause not the message) which I don't have a problem with (and the same goes for `message` not just `cause`), but I'd much rather us tighten the budget on the number of `if` conditions and only use them when information could be lost unless we check for something.

Reply:
@petermetz: Now I got your saying: The `ExceptionHelper.newRuntimeError("Oops", "Doh").cause` was evaluating to `"A CUSTOM EXCEPTION WAS THROWN: Oops - A CUSTOM EXCEPTION WAS THROWN: \"Oops\""` which is exactly what should be prevented -> So the distinction in the else-case made no sense in this scenario you tested yesterday. YES !!

So I did a re-check on your test-scenario with the "Oops", "Doh" and found the root cause why it still showed the duplication: In the function `getExceptionMessageInfo()` in. scenario 3 the thrown exception as in your example is a string only type and was missing `containsCompleteStringifiedException = true` as this scenario handles exception of string only type and there is no other property so it is per se completely stringified. With this fix the distinction in the else case is now working as expected and no longer leading to duplication.